### PR TITLE
proposed test for multi_gpu prediction fault

### LINKED ISSUE #10618: predict with multi-gpu and ddp strategy only get a portion of the predictions
## 🐛 Bug
When called 'predict' method in multi-gpu settings (e.g., 2 gpus, strategy "ddp"), I get only half the predictions (meaning that only the first half of the dataset returns the predictions)

### To Reproduce
I opened a PR with the failing test in 'test_gpu.py"
PR #10619 

### Expected behavior

I expect to get the full predictions for input dataset, potentially on CPU to avoid GPU conflicts
### Additional context
I believe torch.distributed has the right APIs to implement this (if this is really bug). Let me know! :)

cc @justusschock @kaushikb11 @awaelchli @akihironitta

### PR DESCRIPTION
**Pull request by gianscarpe**
## What does this PR do?
Closes #10618

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/helpers/pipelines.py [hunk 1]
```diff
@@ -20,6 +20,16 @@
 from tests.helpers.utils import get_default_logger, load_model_from_checkpoint, reset_seed
 
 
+def run_model_predict(trainer_options: dict, model, data: LightningDataModule):
+    reset_seed()
+    trainer = Trainer(**trainer_options)
+
+    predictions = trainer.predict(model, data.train_dataloader(), return_predictions=True)
+    assert len(predictions) == len(
+        data.train_dataloader()
```
Comment (awaelchli) [inline tests/helpers/pipelines.py:29]: yes, this cannot be satisfied. 

If you run with 2 gpus / processes and one process does predictions on one half and the other on the other half, then in each process you get

assert len(predictions) == len(data.train_dataloader()) / 2

(this assertion will hold if the dataset is evenly divisible by batch size and number of gpus)
